classes/local.php

```diff
@@ -279,26 +279,27 @@ public static function file_is_public(string $contenthash): bool {
 
         $minduplicates = get_config('filter_imageopt', 'minduplicates');
         $publicfilescache = \cache::make('filter_imageopt', 'public_files');
-        $key = 'public_files_' . $minduplicates;
+        $key = 'publicfiles' . '_' . $minduplicates;
 
         if (empty($minduplicates)) {
             return false;
         }
 
-        if (!$publicfilescache->has($key)) {
-            $publicfiles = $DB->get_records_sql(
-                "SELECT contenthash, COUNT(contenthash)
-                   FROM {files}
-                  WHERE filearea <> 'draft'
-                    AND (filearea <> 'public' OR component <> 'filter_imageopt')
-               GROUP BY contenthash
-                 HAVING COUNT(contenthash) >= :count",
-                ['count' => $minduplicates]
+        if (!$publicfiles = $publicfilescache->get($key)) {
+            $publicfilescache->set($key, []);
+        }
+
+        if (!isset($publicfiles[$contenthash])) {
+            $ocurrences = $DB->count_records_select(
+                'files',
+                "contenthash = :contenthash AND filearea <> 'draft' AND (filearea <> 'public' OR component <> 'filter_imageopt')",
+                ['contenthash' => $contenthash]
             );
 
-            $publicfilescache->set($key, array_column($publicfiles, 'contenthash'));
+            $publicfiles[$contenthash] = $ocurrences >= $minduplicates;
         }
 
-        return in_array($contenthash, $publicfilescache->get($key));
+        $publicfilescache->set($key, $publicfiles);
```

Comment:
1) shouldn't this only be set if it has changed?

2) isn't this going to create a single giant record with every file hash in it? If so this won't scale at all

We need to put a lot of thought into the granularity of the cache keys, and I would suggest that the grain size is either at the level of a course or the level of a url. Maybe using the context could also work.

Reply:
I've updated the patch to progressively cache the hashes as individual keys. Note that it looks a bit weird due to a [bug I found](https://tracker.moodle.org/browse/MDL-74438) where you can't retrieve boolean false from `cachestore_file`.

The cache keys are now `publicfiles_[contenthash]_[minduplicates]` and their values are true/false. We could still do it with an array, but I'm not sure if that has much value since we hit the cache once per image request anyway.